In [ ]:
from IPython.core.display import display, HTML

display(HTML('<h1 style="color:#E50914; font-size:100px; font-weight:bold; text-align:center;">NETFLIX</h1>'))


<hr style="border: 3px solid #E50914;">


**Netflix** is one of the most popular media and video streaming platforms. They have over 10,000 movies or TV shows available on their platform. As of mid-2021, they have over 222M subscribers globally.  


## Key Features  

1. **Extensive Content Library** – Offers a vast collection of movies, TV shows, and documentaries across various genres and languages.  

2. **Personalized Recommendations** – Uses AI algorithms to suggest content based on viewing history, ratings, and preferences.  

3. **Ad-Free Streaming** – Provides an uninterrupted viewing experience without advertisements across all paid plans.  

4. **Offline Viewing** – Allows users to download content and watch without an internet connection, ideal for travel.  

5. **Multi-Device Support** – Accessible on smart TVs, laptops, mobile devices, and gaming consoles with seamless cross-device syncing.  


## Problem Statement  

Netflix is aiming to enhance its content strategy by understanding which types of shows and movies will perform best in different regions. The business problem is to analyze Netflix's content data and generate actionable insights that can guide the decision-making process for content production and expansion. The focus is to identify content trends, popular genres, and successful shows/movies to help Netflix grow its business globally.


## Table of Contents  

1. [Introduction 📖](#introduction)  
    1.1. [Problem Statement ❓](#problem-statement)  
    1.2. [About Data 💾](#about-data)  
2. [Data Observation and Exploration 📊](#data-observation-and-exploration)  
    2.1. [Examining Data Types](#examining-data-types)  
    2.2. [Reviewing Sample Values](#reviewing-sample-values)  
    2.3. [Identifying Missing Data](#identifying-missing-data)  
    2.4. [Generating Statistical Summary](#generating-statistical-summary)  
3. [Data Preprocessing 🧹](#data-preprocessing)  
    3.1. [DataType Conversion](#datatype-conversion)  
    3.2. [Exploding Columns](#exploding-columns)  
    3.3. [Handling Missing Values](#handling-missing-values)  
4. [Data Analysis 📈](#data-analysis)  
    4.1. [Non-Graphical Analysis](#non-graphical-analysis)  
    4.2. [Graphical Analysis 📉](#graphical-analysis)  
5. [Conclusion and Future Outlook](#conclusion-and-future-outlook)


## 1. Introduction 📖  
|   |
|---|
| Netflix, one of the world's leading streaming platforms, offers a vast library of movies and TV shows globally. This case study aims to analyze Netflix's content data to derive insights that can help optimize their content strategy and expand globally. |


## 1.1 Problem Statement ❓

### Overview 📖
This analysis aims to explore key factors such as content availability, release trends, and optimal launch timing to help Netflix enhance its strategy.

### Business Problem 💼
Netflix faces several challenges in content strategy, and this analysis aims to answer key questions that will help improve their offerings:
- **Content Availability**: Which types of content are available in different countries?
- **Release Trend**: How has the number of releases changed over the years?
- **Content Comparison**: What are the differences between movies and TV shows on Netflix?
- **Optimal Launch Timing**: When is the best time to launch a new show or movie?
- **Content Focus**: Should Netflix focus more on TV shows or movies in the coming years?

### Expected Outcome 🎯
The expected outcome of this analysis is to provide Netflix with actionable insights based on data trends. These insights will help in making data-driven decisions regarding the production of content, marketing strategies, and expansion into new regions.


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
!curl -o netflix.csv https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/000/940/original/netflix.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 3319k  100 3319k    0     0  14.9M      0 --:--:-- --:--:-- --:--:-- 15.0M


In [ ]:
data=pd.read_csv('netflix.csv')

In [ ]:
data.shape

(8807, 12)

In [ ]:
data.duplicated().sum()

0

## About Data 💾

### Table of Columns
| Column Name    | Description                                           |
|----------------|-------------------------------------------------------|
| **Show_id**     | Unique identifier for each movie or TV show.          |
| **Type**        | Type of content: Movie or TV Show.                    |
| **Title**       | Title of the movie or TV show.                        |
| **Director**    | Director(s) of the movie or TV show.                  |
| **Cast**        | List of actors/actresses in the movie or TV show.     |
| **Country**     | Country where the movie/show was produced.            |
| **Date_added**  | Date the content was added to Netflix.                |
| **Release_year**| Year the movie/show was originally released.         |
| **Rating**      | The TV or movie rating (e.g., PG, R).                 |
| **Duration**    | Duration in minutes (for movies) or number of seasons (for TV shows). |
| **Listed_in**   | Genre of the content (e.g., Comedy, Drama).          |
| **Description** | Brief summary of the movie or TV show.                |

### Overview 📖
This dataset consists of 12 columns, each providing valuable information about the movies and TV shows available on Netflix. Here’s a deeper look into each column:

- **Show_id**: This column contains a unique identifier for each piece of content, which allows us to track and reference specific movies or TV shows across the dataset.
  
- **Type**: This tells us whether the content is a "Movie" or a "TV Show". It helps in distinguishing between different types of content and allows for analysis of trends between the two.

- **Title**: The title of the movie or TV show is a key attribute that will be analyzed for patterns, popularity, and genre classification.

- **Director**: This column lists the director(s) of the movie or TV show. Analyzing this field can help identify which directors have a significant presence on Netflix and if certain directors are associated with high-performing content.

- **Cast**: The list of actors or actresses involved in the movie/show. This is important for understanding which stars contribute most to Netflix's content library, and if content featuring certain actors performs better.

- **Country**: The country in which the movie/show was produced. This is valuable for understanding Netflix's global reach and what content is available in different countries, which could influence regional preferences and content localization strategies.

- **Date_added**: The date the content was added to Netflix. This helps to identify trends over time, such as the frequency of content being added and whether there are particular seasons when Netflix tends to add more content.

- **Release_year**: This column provides the original release year of the movie or TV show. Analyzing this field can show us how Netflix has shifted its content offering over the years, whether there’s a preference for newer vs older content, or if older content is being rediscovered.

- **Rating**: The TV or movie rating (e.g., PG, R, TV-MA). This column helps in understanding Netflix's audience demographic and can be analyzed to identify trends in content ratings over time or by region.

- **Duration**: For movies, this is the total runtime in minutes. For TV shows, it's the number of seasons. Analyzing duration can reveal patterns in the length of content being added to Netflix and which types of content (movies or TV shows) are more popular with longer durations.

- **Listed_in**: The genre(s) of the content. Genres like Comedy, Drama, Action, etc., are essential for identifying trends in content types and preferences among users.

- **Description**: This is a short summary of what the movie/show is about. It helps to understand the type of content Netflix offers and may help in analyzing the success or appeal of certain types of storylines or themes.

### Number of Rows
The dataset contains **8807 rows**. Each row represents a unique movie or TV show listed on Netflix, and the data includes no duplicate rows, meaning every row is a unique entry with distinct content details.

By exploring these columns, we can uncover insights about Netflix's content strategy, understand audience preferences, and make data-driven recommendations for future content development and strategy.


## 2. Data Observation and Exploration 📊

**Examining Data Types:** The dataset contains various types of data, including categorical variables like **Type**, **Rating**, and **Country**, and continuous variables like **Duration** and **Release_year**. These different data types will help in determining which analyses to apply.  

**Reviewing Sample Values:** A quick review of the first few records (head of the dataset) shows that all columns are populated, with no missing data in the sample rows. We will later confirm this by checking the entire dataset.  

**Identifying Missing Data:** There are no missing values in the dataset for the columns we reviewed in the sample. However, a more thorough check will be performed to ensure that the entire dataset is complete.  

**Generating Statistical Summary:** A statistical summary will be generated to analyze the distribution of continuous variables like **Release_year** and **Duration**. This will help us identify outliers or unusual values in the dataset.  

These observations form the foundation for further data exploration, where we will analyze trends and relationships between different features.  


### 2.1 Examining Data Types 🏷️  

We will identify the data type of each column to ensure it is appropriate for analysis. Understanding data types helps in selecting the right preprocessing steps and performing relevant analyses effectively. This will also help in detecting any inconsistencies in the dataset.

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


### Insights from Data Types 🔍  

The dataset contains a mix of **object (string)** and **integer** data types. Before starting the analysis, we might need to perform some type conversions to ensure accurate computations. For example, the **Date_added** column can be converted to a proper **datetime** format, which will allow us to analyze trends over time more effectively.


### 2.2 Reviewing Sample Values 🔍  

We will examine a few rows of the dataset to get an overview of how the data is structured. This helps in understanding the format, completeness, and possible inconsistencies in the dataset before proceeding with further analysis.


In [ ]:
data.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [ ]:
data.tail()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
8802,s8803,Movie,Zodiac,David Fincher,"Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...",United States,"November 20, 2019",2007,R,158 min,"Cult Movies, Dramas, Thrillers","A political cartoonist, a crime reporter and a..."
8803,s8804,TV Show,Zombie Dumb,NaN,NaN,NaN,"July 1, 2019",2018,TV-Y7,2 Seasons,"Kids' TV, Korean TV Shows, TV Comedies","While living alone in a spooky town, a young g..."
8804,s8805,Movie,Zombieland,Ruben Fleischer,"Jesse Eisenberg, Woody Harrelson, Emma Stone, ...",United States,"November 1, 2019",2009,R,88 min,"Comedies, Horror Movies",Looking to survive in a world taken over by zo...
8805,s8806,Movie,Zoom,Peter Hewitt,"Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...",United States,"January 11, 2020",2006,PG,88 min,"Children & Family Movies, Comedies","Dragged from civilian life, a former superhero..."
8806,s8807,Movie,Zubaan,Mozez Singh,"Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...",India,"March 2, 2019",2015,TV-14,111 min,"Dramas, International Movies, Music & Musicals",A scrappy but poor boy worms his way into a ty...


### Insights from Sample Data 💡  

- The dataset contains a mix of movies and TV shows. The **type** column indicates whether the content is a "Movie" or "TV Show."  
- The **cast** and **director** columns have some missing values, indicating incomplete information for certain entries.  
- The **country** column also has missing values, suggesting that the production location is not available for all titles.  
- The **date_added** column is formatted as strings but can be converted to a datetime format for better analysis of trends over time.  
- The **duration** column shows values in different formats, such as "90 min" for movies and "2 Seasons" for TV shows, requiring preprocessing for consistency.  
- The **listed_in** column contains genres, often with multiple genres listed for a single title, which can be exploded for more granular analysis.  
- The **description** column provides a short summary of each title, which might help with sentiment or content-based analysis.  

These observations indicate that while the dataset is rich in information, it requires preprocessing and cleaning to handle missing data and ensure consistent formatting for effective analysis.


### 2.3 Identify Missing Data 🕳️  

We will analyze the dataset to identify missing values in each column. This step is crucial to ensure data completeness and reliability for further analysis. Columns such as **cast**, **director**, and **country** already show some missing values in the sample data, and we will quantify these missing entries across the entire dataset.  

Handling these missing values appropriately will be necessary, either by imputing them with relevant data, removing the rows/columns, or marking them as unavailable for analysis.


In [ ]:
# Calculate the total number of missing values for each column in the dataset
missing_values = data.isna().sum()

# Filter out columns that do not have any missing values
missing_values = missing_values[missing_values != 0]

# Display the columns with missing values and the count of missing entries
missing_values


,0
director,2634
cast,825
country,831
date_added,10
rating,4
duration,3


### Insights from Missing Date 💡  

- The **director** column has the highest number of missing values (2,634 entries), suggesting incomplete data for many titles regarding their directors.  
- The **cast** column is missing 825 entries, indicating that actor details are not available for some titles.  
- The **country** column is missing 831 entries, which might make it challenging to analyze the geographic distribution of content.  
- The **date_added** column has only 10 missing values, which can be addressed with minimal effort to ensure accurate trend analysis.  
- The **rating** column has 4 missing values, which might require handling based on the specific analysis requirements.  
- The **duration** column has 3 missing values, which could be critical for analyzing the length of movies or TV shows.  

These missing values highlight the need for effective preprocessing strategies, such as imputing, excluding, or categorizing missing data, to maintain the quality of the analysis.


## 2.4 Generating Statistical Summary 📊

To gain insights into the dataset, we generate a statistical summary of both numerical and categorical variables.

### 📌 Summary of Numerical Variables
The dataset's numerical columns, such as **Release_year** and **Duration**, are analyzed using descriptive statistics. This includes:
- **Count**: Total number of entries.
- **Mean**: The average value.
- **Standard Deviation**: Measures variability.
- **Minimum and Maximum Values**: Identifies the range of values.
- **Quartiles (25%, 50%, 75%)**: Helps in understanding data distribution.

### 📌 Summary of Categorical Variables
For categorical columns like **Type**, **Rating**, and **Country**, descriptive statistics include:
- **Count**: Number of non-null values.
- **Unique Values**: Distinct categories available.
- **Top Category**: The most frequently occurring value.
- **Frequency**: Count of the most common category.

This statistical summary aids in detecting patterns, identifying anomalies, and preparing data for further analysis.


In [ ]:
data.describe(include='all')

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
count,8807,8807,8807,6173,7982,7976,8797,8807.000000,8803,8804,8807,8807
unique,8807,2,8807,4528,7692,748,1767,NaN,17,220,514,8775
top,s1,Movie,Dick Johnson Is Dead,Rajiv Chilaka,David Attenborough,United States,"January 1, 2020",NaN,TV-MA,1 Season,"Dramas, International Movies","Paranormal activity at a lush, abandoned prope..."
freq,1,6131,1,19,19,2818,109,NaN,3207,1793,362,4
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2014.180198,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,8.819312,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1925.000000,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2013.000000,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2017.000000,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2019.000000,NaN,NaN,NaN,NaN


## 2.4 Insights from Statistical Summary 📊

Based on the statistical summary, the following observations can be made:

### 🔹 Numerical Variables:
- **Release Year**: The dataset spans from **1925 to 2021**, with a mean release year of **2014**, indicating that most of the content is relatively recent.
- **Duration**: Since the duration column is treated as an object (categorical), further cleaning is needed to separate movies (minutes) and TV shows (seasons) for numerical analysis.

### 🔹 Categorical Variables:
- **Type**: The dataset consists of two content types: **Movies and TV Shows**, with Movies being the dominant type (**6,131 entries**).
- **Rating**: The most frequent rating is **TV-MA**, appearing **3,207 times**, suggesting a significant portion of content is targeted at mature audiences.
- **Country**: The most common production country is the **United States**, with **2,818** entries.
- **Director & Cast**: The most frequently credited director is **Rajiv Chilaka (19 titles)**, and the most common actor is **David Attenborough (19 appearances)**.
- **Genres (Listed In)**: The most frequent genre category is **Dramas & International Movies**, which appears **362 times**.

### 🔹 Data Quality Observations:
- Some categorical columns like **Duration** need data transformation for better numerical analysis.
- The **Director and Cast columns** have missing values, indicating incomplete metadata for some entries.
- The **Date Added** field has **1,767 unique values**, meaning content has been added over an extended period, but further analysis is needed to understand trends.

These insights will guide further analysis, such as trends in content release, genre popularity, and country-specific content distributions. 🚀


## Data Preprocessing 🧹
---

We will perform the following data preprocessing operations:

1. **DataType Conversion**  
2. **Exploding Columns**  
3. **Handling Missing Values**  




### 3.1 Data Conversion

Data type conversion is a crucial step in data preprocessing. It ensures that each column is in an appropriate format, making data analysis more efficient and accurate. Incorrect data types can lead to unexpected behavior during computations, inefficient storage usage, and complications when performing operations such as filtering or grouping. Below is the table describing the necessary data type conversions:


| SL No. | **Column Name**   | **Current Data Type** | **Desired Data Type** | **Reason for Conversion** |
|--------|--------------------|------------------------|-----------------------|---------------------------|
| 1      | **`show_id`**      | `object`               | `string`              | ID values should remain as strings to prevent unwanted numerical operations. |
| 2      | **`type`**         | `object`               | `category`            | Limited unique values; converting to category reduces memory usage. |
| 3      | **`title`**        | `object`               | `string`              | Textual data should remain as a string for analysis purposes. |
| 4      | **`director`**     | `object`               | `string`              | Names of directors should remain strings, though missing values need handling. |
| 5      | **`cast`**         | `object`               | `string`              | Similar to `director`, names should remain strings; missing values require handling. |
| 6      | **`country`**      | `object`               | `category`            | Limited unique values; converting to category optimizes storage. |
| 7      | **`date_added`**   | `object`               | `datetime`            | Dates should be in datetime format for easy manipulation and analysis. |
| 8      | **`release_year`** | `int64`                | `int`                 | Year values are integers, so no change is required. |
| 9      | **`rating`**       | `object`               | `category`            | Limited unique values; converting to category improves efficiency. |
| 10     | **`duration`**     | `object`               | `int (minutes)`       | Converting to integer allows for numerical analysis on duration. |
| 11     | **`listed_in`**    | `object`               | `string`              | Genre categories should remain as strings. |
| 12     | **`description`**  | `object`               | `string`              | Descriptions should remain as textual data. |


By performing these conversions, the dataset becomes more structured, optimized for analysis, and easier to process in subsequent steps.



In [ ]:
# Convert 'show_id' to string to ensure it is treated as a unique identifier for each row
data['show_id'] = data['show_id'].astype('string')

# Convert 'type' to category for memory optimization and efficient analysis
data['type'] = data['type'].astype('category')

# Convert 'title' to string to preserve the textual data for analysis
data['title'] = data['title'].astype('string')

# Convert 'director' to string to ensure proper handling of textual data
data['director'] = data['director'].astype('string')

# Convert 'cast' to string to handle textual data about actors
data['cast'] = data['cast'].astype('string')

# Convert 'country' to category to optimize memory and enable efficient grouping
data['country'] = data['country'].astype('category')

# Convert 'date_added' to datetime format for time-based analysis
data['date_added'] = pd.to_datetime(data['date_added'].str.strip())

# Convert 'release_year' to integer for numerical computations and analysis
data['release_year'] = data['release_year'].astype('int')

# Convert 'rating' to category for efficient analysis of rating distribution
data['rating'] = data['rating'].astype('category')

# Extract numeric values from 'duration' and convert to integer; handle missing values appropriately
data['duration'] = data['duration'].apply(lambda x: int(x.split()[0]) if not pd.isnull(x) else x)

# Convert 'listed_in' to string to preserve genre information as textual data
data['listed_in'] = data['listed_in'].astype('string')

# Convert 'description' to string to retain summary information about the content
data['description'] = data['description'].astype('string')


In [ ]:
#Glance at data now
print(data.dtypes)
data.head()

show_id         string[python]
type                  category
title           string[python]
director        string[python]
cast            string[python]
country               category
date_added      datetime64[ns]
release_year             int64
rating                category
duration               float64
listed_in       string[python]
description     string[python]
dtype: object


,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,<NA>,United States,2021-09-25,2020,PG-13,90.0,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,<NA>,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,2021-09-24,2021,TV-MA,2.0,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,2021-09-24,2021,TV-MA,1.0,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,<NA>,<NA>,NaN,2021-09-24,2021,TV-MA,1.0,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,<NA>,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,2021-09-24,2021,TV-MA,2.0,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


### 3.2 Exploding Columns 💥  

In the dataset, certain columns like **cast**, **director**, **listed_in**, and **country** contain multiple values separated by commas. To perform detailed analysis, it is beneficial to explode these columns, where each value gets its own row.  

1. **Cast**:  
   - The **cast** column lists multiple actors in a single row.  
   - Exploding this column allows us to analyze the frequency and contributions of individual actors.  

2. **Director**:  
   - The **director** column contains multiple directors for some entries.  
   - By exploding this column, we can evaluate the contributions of each director separately.  

3. **Listed_in**:  
   - This column represents genres for movies and TV shows.  
   - Exploding it enables us to study which genres are most common in the Netflix catalog.  

4. **Country**:  
   - The **country** column lists multiple countries where some content was produced.  
   - Exploding it helps analyze the geographical distribution of content and identify trends across countries.  

This step is crucial for granular insights into the data and will help answer key questions, such as:  
- Which actors and directors contribute the most to Netflix content?  
- What genres dominate Netflix’s catalog?  
- How is content distributed across different countries?  

Exploding these columns allows for more focused and meaningful analysis. Let’s proceed with implementing this transformation! 🚀


In [ ]:
# Split the 'cast' column into a list of actors (by splitting on commas) and explode the list into separate rows
data = data.assign(cast=data.cast.str.split(',')).explode('cast')

# Reset the index after exploding and rename the 'cast' column to 'actor'
data = data.reset_index(drop=True).rename(columns={'cast': 'actor'})

# Remove any leading or trailing whitespace from the 'actor' column
data['actor'] = data['actor'].str.strip()

# Split the 'listed_in' column into a list of genres (by splitting on commas) and explode the list into separate rows
data = data.assign(listed_in=data.listed_in.str.split(',')).explode('listed_in')

# Reset the index after exploding and rename the 'listed_in' column to 'genre'
data = data.reset_index(drop=True).rename(columns={'listed_in': 'genre'})

# Remove any leading or trailing whitespace from the 'genre' column
data['genre'] = data['genre'].str.strip()

# Split the 'director' column into a list of directors (by splitting on commas) and explode the list into separate rows
data = data.assign(director=data.director.str.split(',')).explode('director')

# Remove any leading or trailing whitespace from the 'director' column
data['director'] = data['director'].str.strip()

# Split the 'country' column into a list of countries (by splitting on commas) and explode the list into separate rows
data = data.assign(country=data.country.str.split(',')).explode('country')

# Remove any leading or trailing whitespace from the 'country' column
data['country'] = data['country'].str.strip()

#Display
data.head()

,show_id,type,title,director,actor,country,date_added,release_year,rating,duration,genre,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,<NA>,United States,2021-09-25,2020,PG-13,90.0,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,<NA>,Ama Qamata,South Africa,2021-09-24,2021,TV-MA,2.0,International TV Shows,"After crossing paths at a party, a Cape Town t..."
2,s2,TV Show,Blood & Water,<NA>,Ama Qamata,South Africa,2021-09-24,2021,TV-MA,2.0,TV Dramas,"After crossing paths at a party, a Cape Town t..."
3,s2,TV Show,Blood & Water,<NA>,Ama Qamata,South Africa,2021-09-24,2021,TV-MA,2.0,TV Mysteries,"After crossing paths at a party, a Cape Town t..."
4,s2,TV Show,Blood & Water,<NA>,Khosi Ngema,South Africa,2021-09-24,2021,TV-MA,2.0,International TV Shows,"After crossing paths at a party, a Cape Town t..."


### 3.3 Handling Missing Values 🛠️

In this section, we address the missing values in the dataset to ensure data integrity and reliability for analysis. As identified earlier, the following columns contain missing values:
- **Director**
- **Cast**
- **Country**
- **Date Added**
- **Rating**
- **Duration**

We will use appropriate strategies to handle these missing values based on the data type and context of each column.


In [ ]:
missing_values=data.isna().sum()[data.isna().sum()!=0]
missing_values

,0
director,50643
actor,2149
country,11897
date_added,158
rating,67
duration,3


#### 3.3.1 Handling Missing Values in Director 🎥

To handle the missing values in the **Director** column, we will replace the missing entries with a placeholder value such as "Unknown". This ensures that the missing data does not disrupt the analysis while preserving the overall dataset structure.

In [ ]:
missing_director=data[data['director'].isna()]
show_id_of_missing_director=missing_director['show_id']
missing_director.head()

,show_id,type,title,director,actor,country,date_added,release_year,rating,duration,genre,description
1,s2,TV Show,Blood & Water,<NA>,Ama Qamata,South Africa,2021-09-24,2021,TV-MA,2.0,International TV Shows,"After crossing paths at a party, a Cape Town t..."
2,s2,TV Show,Blood & Water,<NA>,Ama Qamata,South Africa,2021-09-24,2021,TV-MA,2.0,TV Dramas,"After crossing paths at a party, a Cape Town t..."
3,s2,TV Show,Blood & Water,<NA>,Ama Qamata,South Africa,2021-09-24,2021,TV-MA,2.0,TV Mysteries,"After crossing paths at a party, a Cape Town t..."
4,s2,TV Show,Blood & Water,<NA>,Khosi Ngema,South Africa,2021-09-24,2021,TV-MA,2.0,International TV Shows,"After crossing paths at a party, a Cape Town t..."
5,s2,TV Show,Blood & Water,<NA>,Khosi Ngema,South Africa,2021-09-24,2021,TV-MA,2.0,TV Dramas,"After crossing paths at a party, a Cape Town t..."


In [ ]:
# Fill missing values in the 'director' column with the placeholder 'Unknown_director'
data['director'] = data['director'].fillna('Unknown_director')

# Display the rows where the 'show_id' corresponds to entries with missing directors
# (Assuming 'show_id_of_missing_director' is a predefined list of such IDs)
data[data['show_id'].isin(show_id_of_missing_director)].head()

,show_id,type,title,director,actor,country,date_added,release_year,rating,duration,genre,description
1,s2,TV Show,Blood & Water,Unknown_director,Ama Qamata,South Africa,2021-09-24,2021,TV-MA,2.0,International TV Shows,"After crossing paths at a party, a Cape Town t..."
2,s2,TV Show,Blood & Water,Unknown_director,Ama Qamata,South Africa,2021-09-24,2021,TV-MA,2.0,TV Dramas,"After crossing paths at a party, a Cape Town t..."
3,s2,TV Show,Blood & Water,Unknown_director,Ama Qamata,South Africa,2021-09-24,2021,TV-MA,2.0,TV Mysteries,"After crossing paths at a party, a Cape Town t..."
4,s2,TV Show,Blood & Water,Unknown_director,Khosi Ngema,South Africa,2021-09-24,2021,TV-MA,2.0,International TV Shows,"After crossing paths at a party, a Cape Town t..."
5,s2,TV Show,Blood & Water,Unknown_director,Khosi Ngema,South Africa,2021-09-24,2021,TV-MA,2.0,TV Dramas,"After crossing paths at a party, a Cape Town t..."


### 3.3.2 Handling Actor (Cast)

To handle missing values in the `cast` column, we will replace them with **"Unknown Actor."** Leaving these values empty could mislead the analysis, as it may appear that the data is incomplete or incorrectly recorded. Using a placeholder ensures consistency and avoids misinterpretation while performing analysis.


In [ ]:
missing_actor_data=data[data['actor'].isna()]
missing_actor_show_id=missing_actor_data['show_id']

missing_actor_data.head(3)

,show_id,type,title,director,actor,country,date_added,release_year,rating,duration,genre,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,<NA>,United States,2021-09-25,2020,PG-13,90.0,Documentaries,"As her father nears the end of his life, filmm..."
85,s4,TV Show,Jailbirds New Orleans,Unknown_director,<NA>,NaN,2021-09-24,2021,TV-MA,1.0,Docuseries,"Feuds, flirtations and toilet talk go down amo..."
86,s4,TV Show,Jailbirds New Orleans,Unknown_director,<NA>,NaN,2021-09-24,2021,TV-MA,1.0,Reality TV,"Feuds, flirtations and toilet talk go down amo..."


In [ ]:
#fillna for actor with unknown_actor
data['actor']=data['actor'].fillna('unknown_actor')
data[data['show_id'].isin(missing_actor_show_id)].head(3)

,show_id,type,title,director,actor,country,date_added,release_year,rating,duration,genre,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,unknown_actor,United States,2021-09-25,2020,PG-13,90.0,Documentaries,"As her father nears the end of his life, filmm..."
85,s4,TV Show,Jailbirds New Orleans,Unknown_director,unknown_actor,NaN,2021-09-24,2021,TV-MA,1.0,Docuseries,"Feuds, flirtations and toilet talk go down amo..."
86,s4,TV Show,Jailbirds New Orleans,Unknown_director,unknown_actor,NaN,2021-09-24,2021,TV-MA,1.0,Reality TV,"Feuds, flirtations and toilet talk go down amo..."


### 3.3.3 Handling Country

To handle missing values in the `country` column, we will fill them with the **most frequent country.** This approach ensures that the majority pattern in the dataset is maintained and prevents misinterpretation due to missing data.


In [ ]:
missing_country_data=data[data['country'].isna()]
show_id_missing_country=missing_country_data['show_id']

missing_country_data.head(3)

,show_id,type,title,director,actor,country,date_added,release_year,rating,duration,genre,description
58,s3,TV Show,Ganglands,Julien Leclercq,Sami Bouajila,NaN,2021-09-24,2021,TV-MA,1.0,Crime TV Shows,To protect his family from a powerful drug lor...
59,s3,TV Show,Ganglands,Julien Leclercq,Sami Bouajila,NaN,2021-09-24,2021,TV-MA,1.0,International TV Shows,To protect his family from a powerful drug lor...
60,s3,TV Show,Ganglands,Julien Leclercq,Sami Bouajila,NaN,2021-09-24,2021,TV-MA,1.0,TV Action & Adventure,To protect his family from a powerful drug lor...


In [ ]:
unique_country_as_per_show_id=data.groupby('show_id')['country'].unique()

print("***Unique Country as Per Show Id***",unique_country_as_per_show_id)

most_common_country=unique_country_as_per_show_id.explode().dropna().mode()

print("\nMost commen Country",most_common_country[0])

***Unique Country as Per Show Id*** show_id
s1                   [United States]
s10                  [United States]
s100         [France, United States]
s1000       [Germany, United States]
s1001                          [nan]
                    ...             
s995                       [Nigeria]
s996                        [Turkey]
s997                         [Japan]
s998     [Australia, United Kingdom]
s999                         [India]
Name: country, Length: 8807, dtype: object

Most commen Country United States
